### PR TITLE
[PASS] add constant folding pass

### PR DESCRIPTION
**Pull request by yuanlehome**
### PR types

Others
### PR changes

Others
### Description

- 基于新IR新PASS体系新增constant folding pass
- 调整pattern_rewrite_test和pass_manager_test单测实现细节
- 优化phi_kernel_adaptor关于pd.fetch op部分的逻辑
- 调整dce pass的目录以及命名规范
- 其它一些代码细节

TODO:
- scope逻辑去除？
- 尝试其他实现方案，下沉常量折叠功能

### Others
Pcard-71500

### Notes
328: ===-------------------------------------------------------------------------===
328:           IRPrinting on builtin.module before ConstantFoldingPass pass
328: ===-------------------------------------------------------------------------===
328:  () = "builtin.module" () {program:0x7ffd4d4500f0} : () -> 
328: {
328:  (%0) = "pd.full" () {place:Place(cpu),dtype:float32,shape:IntArray[4,3,16,16],value:1.5} : () -> tensor<4x3x16x16xf32>
328:  (%1) = "pd.full" () {place:Place(cpu),dtype:float32,shape:IntArray[64,3,3,3],value:1.5} : () -> tensor<64x3x3x3xf32>
328:  (%2) = "pd.full" () {place:Place(cpu),dtype:float32,shape:IntArray[64],value:1.5} : () -> tensor<64xf32>
328:  (%3) = "pd.full" () {place:Place(cpu),dtype:float32,shape:IntArray[64],value:1.5} : () -> tensor<64xf32>
328:  (%4) = "pd.full" () {place:Place(cpu),dtype:float32,shape:IntArray[64],value:1.5} : () -> tensor<64xf32>
328:  (%5) = "pd.full" () {place:Place(cpu),dtype:float32,shape:IntArray[64],value:1.5} : () -> tensor<64xf32>
328:  (%6) = "pd.full" () {place:Place(cpu),dtype:float32,shape:IntArray[64],value:1e-06} : () -> tensor<64xf32>
328:  (%7) = "pd.add" (%3, %6) {} : (tensor<64xf32>, tensor<64xf32>) -> tensor<64xf32>
328:  (%8) = "pd.sqrt" (%7) {} : (tensor<64xf32>) -> tensor<64xf32>
328:  (%9) = "pd.divide" (%4, %8) {} : (tensor<64xf32>, tensor<64xf32>) -> tensor<64xf32>
328:  (%10) = "pd.full_int_array" () {place:Place(cpu),value:IntArray[64,1,1,1],dtype:int64} : () -> tensor<4xi64>
328:  (%11, %12) = "pd.reshape" (%9, %10) {} : (tensor<64xf32>, tensor<4xi64>) -> tensor<64x1x1x1xf32>, tensor<0x64xf32>
328:  (%13) = "pd.multiply" (%1, %11) {} : (tensor<64x3x3x3xf32>, tensor<64x1x1x1xf32>) -> tensor<64x3x3x3xf32>
328:  (%14) = "pd.conv2d" (%0, %13) {groups:1,padding_algorithm:EXPLICIT,strides:array[1,1],data_format:NCHW,dilations:array[1,1],paddings:array[0,0]} : (tensor<4x3x16x16xf32>, tensor<64x3x3x3xf32>) -> tensor<4x64x14x14xf32>
328:  (%15) = "pd.multiply" (%2, %9) {} : (tensor<64xf32>, tensor<64xf32>) -> tensor<64xf32>
328:  (%16) = "pd.subtract" (%5, %15) {} : (tensor<64xf32>, tensor<64xf32>) -> tensor<64xf32>
328:  (%17) = "pd.full_int_array" () {place:Place(cpu),value:IntArray[1,64,1,1],dtype:int64} : () -> tensor<4xi64>
328:  (%18, %19) = "pd.reshape" (%16, %17) {} : (tensor<64xf32>, tensor<4xi64>) -> tensor<1x64x1x1xf32>, tensor<0x64xf32>
328:  (%20) = "pd.add" (%14, %18) {} : (tensor<4x64x14x14xf32>, tensor<1x64x1x1xf32>) -> tensor<4x64x14x14xf32>
328:  (%21) = "pd.transpose" (%20) {perm:array[0,2,3,1]} : (tensor<4x64x14x14xf32>) -> tensor<4x14x14x64xf32>
328:  (%22) = "pd.transpose" (%20) {perm:array[0,1,2,3]} : (tensor<4x64x14x14xf32>) -> tensor<4x64x14x14xf32>
328:  (%23) = "pd.fetch" (%22) {col:0,name:out} : (tensor<4x64x14x14xf32>) -> tensor<4x64x14x14xf32>
328: }
328: 
328: 
328: ===-------------------------------------------------------------------------===
328:           IRPrinting on builtin.module after ConstantFoldingPass pass
328: ===-------------------------------------------------------------------------===
328:  () = "builtin.module" () {program:0x7ffd4d4500f0} : () -> 
328: {
328:  (%0) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_1} : () -> tensor<4x3x16x16xf32>
328:  (%1) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_3} : () -> tensor<64x3x3x3xf32>
328:  (%2) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_5} : () -> tensor<64xf32>
328:  (%3) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_7} : () -> tensor<64xf32>
328:  (%4) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_9} : () -> tensor<64xf32>
328:  (%5) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_11} : () -> tensor<64xf32>
328:  (%6) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_13} : () -> tensor<64xf32>
328:  (%7) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_15} : () -> tensor<64xf32>
328:  (%8) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_17} : () -> tensor<64xf32>
328:  (%9) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_19} : () -> tensor<64xf32>
328:  (%10) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_21} : () -> tensor<4xi64>
328:  (%11) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_23} : () -> tensor<64x1x1x1xf32>
328:  (%12) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_25} : () -> tensor<64x3x3x3xf32>
328:  (%13) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_27} : () -> tensor<4x64x14x14xf32>
328:  (%14) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_29} : () -> tensor<64xf32>
328:  (%15) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_31} : () -> tensor<64xf32>
328:  (%16) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_33} : () -> tensor<4xi64>
328:  (%17) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_35} : () -> tensor<1x64x1x1xf32>
328:  (%18) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_37} : () -> tensor<4x64x14x14xf32>
328:  (%19) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_39} : () -> tensor<4x14x14x64xf32>
328:  (%20) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_41} : () -> tensor<4x64x14x14xf32>
328:  (%21) = "pd.fetch" (%20) {col:0,name:out} : (tensor<4x64x14x14xf32>) -> tensor<4x64x14x14xf32>
328: }
328: 
328: ===-------------------------------------------------------------------------===
328:         IRPrinting on builtin.module after DeadCodeEliminationPass pass
328: ===-------------------------------------------------------------------------===
328:  () = "builtin.module" () {program:0x7ffd4d4500f0} : () -> 
328: {
328:  (%0) = "builtin.get_parameter" () {parameter_name:constant_folding_pass_41} : () -> tensor<4x64x14x14xf32>
328:  (%1) = "pd.fetch" (%0) {col:0,name:out} : (tensor<4x64x14x14xf32>) -> tensor<4x64x14x14xf32>
328: }
328:

### CODE REVIEW

File: paddle/fluid/ir/transforms/constant_folding_pass.cc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#include "paddle/ir/pattern_rewrite/frozen_rewrite_pattern_set.h"
+#include "paddle/ir/pattern_rewrite/pattern_match.h"
+#include "paddle/ir/pattern_rewrite/pattern_rewrite_driver.h"
+
+namespace {
+
+class ConstantFoldingPattern : public ir::RewritePattern {
+ public:
+  ConstantFoldingPattern(ir::IrContext* context,
+                         ir::PatternBenefit benefit = 1,
```
Comment (jiweibo) [inline paddle/fluid/ir/transforms/constant_folding_pass.cc:49]: 最后处理常量，先处理常量问题应该不大，可设为1

File: paddle/ir/transforms/dce.cc [hunk 2]
```diff
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 #include "paddle/ir/transforms/dce.h"
-#include <memory>
```
Comment (jiweibo) [inline paddle/ir/transforms/dce.cc:16]: 不是用了unique_ptr吗

File: test/cpp/ir/pattern_rewrite/pattern_rewrite_test.cc [hunk 3]
```diff
@@ -332,21 +333,43 @@ class Conv2dBnFusePattern
...
-    ps.Add<Conv2dBnFusePattern>(op->ir_context());
 
-    ir::FrozenRewritePatternSet frozen_ps(std::move(ps));
+  bool Initialize(ir::IrContext *context) override {
+    ir::RewritePatternSet ps(context);
+    ps.Add<RedundantTransposeFusePattern>(context);
+    auto conv_bn_pattern = std::make_unique<Conv2dBnFusePattern>(
+        context,
+        1,
+        std::vector<std::string>{paddle::dialect::FullOp::name(),
```
Comment (jiweibo) [inline test/cpp/ir/pattern_rewrite/pattern_rewrite_test.cc:978]: 这个好像没有用到的地方，未来提供示例的时候可以不加

**Review by jiweibo — DISMISSED** [review]
LGTM

File: paddle/ir/transforms/dce.cc (hunk 2 above)
Comment (yuanlehome) [reply paddle/ir/transforms/dce.cc:16]: .h里包含了

File: test/cpp/ir/pattern_rewrite/pattern_rewrite_test.cc (hunk 3 above)
Comment (yuanlehome) [reply test/cpp/ir/pattern_rewrite/pattern_rewrite_test.cc:978]: 作为单测 or demo，既然提供了这样的功能，可以加一下使用示例～

File: paddle/fluid/ir/transforms/constant_folding_pass.cc [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    // prepare op outputs
+    std::vector<ir::Type> output_types;
+    for (uint32_t i = 0; i < op->num_results(); i++) {
+      output_types.push_back(op->result(i).type());
+    }
+
+    auto* temp_op =
+        builder.Build(op_inputs, op->attributes(), output_types, op->info());
+
+    // TODO(liuyuanle): Support multiple output.
```
Comment (yuanlehome) [inline paddle/fluid/ir/transforms/constant_folding_pass.cc:145]: 这一块需要等待新ir执行器的bug fix ready.

File: paddle/fluid/ir/transforms/constant_folding_pass.cc [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    builder.Build<paddle::dialect::FetchOp>(
+        temp_op->result(0), "fetch_" + std::to_string(suffix_++), 0);
+    // }
+
+    return program;
+  }
+
+ private:
+  static size_t suffix_;
+  static paddle::framework::Scope scope_;
```
Comment (yuanlehome) [inline paddle/fluid/ir/transforms/constant_folding_pass.cc:162]: 这里涉及scope的逻辑，之后如何修改？ir/pass分析阶段如何与权重以及权重背后的tensor打交道？需要等待训练侧关于scope和parameter打通的设计方案ready。

File: paddle/fluid/ir/transforms/constant_folding_pass.cc [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+    // TODO(liuyuanle): Support multiple output.
+    auto out_tensor = PADDLE_GET_CONST(phi::DenseTensor, fetch_list[0]);
+    std::unique_ptr<ir::Parameter> parameter = std::make_unique<ir::Parameter>(
+        reinterpret_cast<void*>(out_tensor.data()),
+        out_tensor.numel() * phi::SizeOf(out_tensor.dtype()),
+        op->result(0).type());
+
+    std::string param_name =
+        "constant_folding_pass_" + std::to_string(suffix_++);
```
Comment (jiweibo) [inline paddle/fluid/ir/transforms/constant_folding_pass.cc]: 搞一个更特殊的名字 加一些 # @ $这样的名字？

File: paddle/fluid/ir/transforms/constant_folding_pass.cc [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    auto* param_var = scope_.Var(param_name);
+    auto* param_tensor = param_var->GetMutable<phi::DenseTensor>();
+    *param_tensor = out_tensor;
+    program->SetParameter(param_name, std::move(parameter));
+    // rewriter.SetInsertionPoint(op);
+    auto get_parameter_op =
+        rewriter.Build<ir::GetParameterOp>(param_name, op->result(0).type());
+
+    rewriter.ReplaceAllUsesWith(op->result(0), get_parameter_op->result(0));
+    rewriter.EraseOp(op);
```
Comment (jiweibo) [inline paddle/fluid/ir/transforms/constant_folding_pass.cc:104]: 旧的parameter也对应删除了吗

File: paddle/ir/CMakeLists.txt [hunk 8]
```diff
@@ -37,7 +37,7 @@ endfunction()
 add_subdirectory(core)
 add_subdirectory(pass)
 add_subdirectory(pattern_rewrite)
-add_subdirectory(transforms)
+add_subdirectory(builtin_transforms)
```
Comment (jiweibo) [inline paddle/ir/CMakeLists.txt:40]: 旧名字是不是更好些

File: paddle/fluid/ir/transforms/constant_folding_pass.cc (hunk 6 above)
Comment (yuanlehome) [reply paddle/fluid/ir/transforms/constant_folding_pass.cc]: 可以的

File: paddle/fluid/ir/transforms/constant_folding_pass.cc (hunk 7 above)
Comment (yuanlehome) [reply paddle/fluid/ir/transforms/constant_folding_pass.cc:104]: 暂时没有，我看一下

File: paddle/ir/CMakeLists.txt (hunk 8 above)
Comment (yuanlehome) [reply paddle/ir/CMakeLists.txt:40]: 同名的话感觉会和fluid下的引起误解

File: paddle/fluid/ir/transforms/constant_folding_pass.cc (hunk 7 above)
Comment (yuanlehome) [reply paddle/fluid/ir/transforms/constant_folding_pass.cc:104]: 改为通过dce pass来删

File: test/cpp/ir/pattern_rewrite/CMakeLists.txt [hunk 9]
```diff
@@ -1,9 +1,9 @@
...
-  DEPS
-  ir
-  pd_dialect
-  transform_general_functions
-  gtest)
+set(PATTERN_REWRITE_TEST_DEPS _constant_folding_pass
+                              transform_general_functions gtest pd_dialect ir)
+
+if(WITH_DISTRIBUTE)
+  set(PATTERN_REWRITE_TEST_DEPS ${PATTERN_REWRITE_TEST_DEPS} fleet_executor)
```
Comment (jiweibo) [inline test/cpp/ir/pattern_rewrite/CMakeLists.txt:5]: 间接依赖了fleet_executor？

**Review by jiweibo — APPROVED** [review]
LGTM

**Pull request merged**